src/comgr.cpp

```diff
@@ -0,0 +1,490 @@
+
+    public:
+    Dataset() { EC_THROW(amd_comgr_create_data_set(&handle)); }
+    ~Dataset() { EC(amd_comgr_destroy_data_set(handle)); }
+    auto operator()() const { return handle; }
+    void AddData(const Data& d) const { EC_THROW(amd_comgr_data_set_add(handle, d())); }
+    void AddData(const std::string& name,
+                 const std::string& content,
+                 const amd_comgr_data_kind_t type) const
+    {
+        const Data d(type);
+        MIOPEN_LOG_I2(name << ' ' << content.size() << " bytes");
+        d.SetName(name);
+        d.SetBytes(content);
+        AddData(d);
+        const auto show_first = miopen::Value(MIOPEN_DEBUG_COMGR_LOG_SOURCE_TEXT{}, 0);
+        if(show_first > 0 && miopen::IsLogging(miopen::LoggingLevel::Info) &&
+           type == AMD_COMGR_DATA_KIND_SOURCE)
+        {
+            const auto text_length = (content.size() > show_first) ? show_first : content.size();
+            const std::string text(content, 0, text_length);
+            MIOPEN_LOG_I(text);
+        }
+    }
+    size_t GetDataCount(const amd_comgr_data_kind_t kind) const
+    {
+        std::size_t count = 0;
+        ECI_THROW(amd_comgr_action_data_count(handle, kind, &count), kind);
+        return count;
+    }
+    Data GetData(const amd_comgr_data_kind_t kind, const std::size_t index) const
+    {
+        amd_comgr_data_t d;
+        ECI_THROW(amd_comgr_action_data_get_data(handle, kind, index, &d), kind);
+        return {d};
+    }
+};
+
+class ActionInfo : ComgrOwner
+{
+    amd_comgr_action_info_t handle = {0};
+
+    public:
+    ActionInfo() { EC_THROW(amd_comgr_create_action_info(&handle)); }
+    ~ActionInfo() { EC(amd_comgr_destroy_action_info(handle)); }
+    auto operator()() const { return handle; }
+    void SetLanguage(const amd_comgr_language_t language) const
+    {
+        ECI_THROW(amd_comgr_action_info_set_language(handle, language), language);
+    }
+    void SetIsaName(const std::string& isa) const
+    {
+        ECI_THROW_MSG(amd_comgr_action_info_set_isa_name(handle, isa.c_str()), isa, isa);
+    }
+    void SetLogging(const bool state) const
+    {
+        ECI_THROW(amd_comgr_action_info_set_logging(handle, state), state);
+    }
+    void SetOptionList(const std::vector<std::string>& options) const
+    {
+        std::vector<const char*> vp;
+        vp.reserve(options.size());
+        for(auto& opt : options) // cppcheck-suppress useStlAlgorithm
+            vp.push_back(opt.c_str());
+        LogOptions(vp.data(), vp.size());
+        ECI_THROW(amd_comgr_action_info_set_option_list(handle, vp.data(), vp.size()), vp.size());
+    }
+    void Do(const amd_comgr_action_kind_t kind, const Dataset& in, const Dataset& out) const
+    {
+        ECI_THROW_MSG(amd_comgr_do_action(kind, handle, in(), out()), kind, GetLog(out, true));
+        const auto log = GetLog(out);
+        if(!log.empty())
+            MIOPEN_LOG_I(to_string(kind) << ": " << GetLog(out));
```

Comment:
It seems like you call `GetLog` three times for the same dataset. It seems unnecessary. Shouldn't this be `MIOPEN_LOG_I(to_string(kind) << ": " << log)`?

Reply:
Hoho. Sure.

Reply:
> It seems like you call GetLog three times for the same dataset. 

Actually twice. The first call (line 375) only happens just prior throwing, so lines 376-377 won't be reached.

Reply:
Why does it need to be called twice? Why will the log be different the second time? Its hard to tell what happening in the code with the `ECI_THROW_MSG` macros.

Reply:
now it is called once

Reply:
it was twice when it was
```
        const auto log = GetLog(out);
        if(!log.empty())
            MIOPEN_LOG_I(to_string(kind) << ": " << GetLog(out));
```
now it ends with `...<< log);`

Reply:
Its still called twice here:

```cpp
        ECI_THROW_MSG(amd_comgr_do_action(kind, handle, in(), out()), kind, GetLog(out, true));
        const auto log = GetLog(out);
```

Why cant `log` use the same log from the `ECI_THROW_MSG` macro?

Reply:
Because 
- `GetLog()` is invoked with different parameters, see https://github.com/ROCmSoftwarePlatform/MIOpen/pull/177#discussion_r418628804
- In the macro, `GetLog()` is invoked only just prior throwing. If control flow returns from `ECI_THROW_MSG`, then `GetLog()` wasn't called there.

Perhaps I will get rid of `comgr_error_handling` parameter in `GetLog()` and unify the code. Right now I think it is too soon to decide. Nevertheless `GetLog()` is invoked only once.

Reply:
Can we get rid of the macros? It makes hard to see that because it looks like `GetLog(out, true)` is called before `amd_comgr_do_action`.

Reply:
Please see https://github.com/ROCmSoftwarePlatform/MIOpen/pull/177#discussion_r417411528